### PR TITLE
[Feature] include/exclude filters (#1)

### PR DESCRIPTION
**Pull request by rnsc**
Hello @Ana06 ,

Here's my attempt at the include/exclude filtering following issue #1.
I've picked up the suggestion regarding the use of regex from @micalevisk.

Disclaimer, I'm no programmer, so I hope I haven't made any terrible mistakes.

### CODE REVIEW

**Review by micalevisk — APPROVED** [review]
well done!

**Review by micalevisk — APPROVED** [review]

**Comment by rnsc** [comment]
@micalevisk @Ana06 do you think we should already merge this as is and review the suggestion on the issue #1 first or rework this completely?

**Comment by micalevisk** [comment]
> @micalevisk @Ana06 do you think we should already merge this as is and review the suggestion on the issue #1 first or rework this completely?

I like [that suggestion](https://github.com/Ana06/get-changed-files/issues/1#issuecomment-870558061) because it seems like more idiomatic to GH Actions

```ts
with:
  paths:
    - '*.yml'
    - '!blah/test/**'
```

To implement this, I'd say is better to open a new PR and close this one.

File: .github/workflows/test.yml
Comment (Ana06) [inline .github/workflows/test.yml:4]: why is this needed? 🤔

File: .github/workflows/test.yml
Comment (Ana06) [inline .github/workflows/test.yml:92]: [nitpick] I think the name was accurate, but it is not very important. Does someone else has an opinion about this?

File: dist/index.js
Comment (Ana06) [inline dist/index.js:3574]: I think we should add some documentation about how this work. For example, what happens when you have the same file in both include and exclude. Just so that people find this information on the README instead of needing to check the code.

**Review by Ana06 — COMMENTED** [review]
Great work @rnsc! Thank you so much for sending a PR ❤️ and sorry for the late review, busy week! 😅 
Some documentation about this great new features on the README would be nice 😉 

Thanks @micalevisk for reviewing the PR and for sharing your opinion! 😄 

I really do not have an strong opinion about what option we should take (GH Actions like vs include/exclude). As already two persons mentioned they prefer the GH Actions like syntax I tend to think we should use that one. @rnsc do you want to work on that or do you prefer we merge this (maybe after addressing the minor changes and adding documentation if you have time) and then someone else sends a PR for the GH Actions like implementation?

File: .github/workflows/test.yml
Comment (rnsc) [reply .github/workflows/test.yml:4]: Good question, while testing, it was handy for me to just trigger a build without a PR or a push, I'll remove it!

File: .github/workflows/test.yml
Comment (rnsc) [reply .github/workflows/test.yml:92]: I wanted to have a different name since I had duplicated the step with the exclude filter.

File: dist/index.js
Comment (rnsc) [reply dist/index.js:3574]: Yes you're right, I totally zapped the README update and it's indeed very much needed, I'll work on that tomorrow or Friday.

**Comment by rnsc** [comment]
> Great work @rnsc! Thank you so much for sending a PR ❤️ and sorry for the late review, busy week! 😅
> Some documentation about this great new features on the README would be nice 😉
> 
> Thanks @micalevisk for reviewing the PR and for sharing your opinion! 😄
> 
> I really do not have an strong opinion about what option we should take (GH Actions like vs include/exclude). As already two persons mentioned they prefer the GH Actions like syntax I tend to think we should use that one. @rnsc do you want to work on that or do you prefer we merge this (maybe after addressing the minor changes and adding documentation if you have time) and then someone else sends a PR for the GH Actions like implementation?

Hey @Ana06 Thank you for taking the time to review this, don't feel pressured to do so if you have other stuff on your hands though, I was merely mentioning you for visibility regarding the feedback I had received on the issue for the feature request itself!


Adding  @agraul @micalevisk too for the following inquiry:

I'll check if I can do something about accepting a list of containing both include and exclude like suggested in this example: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-positive-and-negative-patterns-1

The paths definition for files is not a regex but a glob pattern though.
Do you think the use of a library like this one [glob](https://www.npmjs.com/package/glob) is ok?

**Comment by micalevisk** [comment]
> Do you think the use of a library like this one glob is ok?

I think it is since it follows the [filter pattern cheat sheet](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet) _(you could add this cheat sheet to the README)_

**Comment by agraul** [comment]
> The paths definition for files is not a regex but a glob pattern though.

Isn't that already the case? Looking at the [README](https://github.com/Ana06/get-changed-files#usage) and the [current code](), the default filter is `*`, which is a [quantifier](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Quantifiers) and doesn't match anything by itself. I don't understand how it currently works, JS is not a language I know well. I would expect  `.*` is needed to match everything.

> Do you think the use of a library like this one [glob](https://www.npmjs.com/package/glob) is ok?

I'm no expert, from a first glance it looks like it should work.

**Comment by rnsc** [comment]
Hello @Ana06 @micalevisk @agraul ,

I believe I have the correct logic in place to support glob filtering.
I've updated the README with the example.
Should I also mention the library used in the README? <https://github.com/isaacs/minimatch>
So that people understand how the glob pattern matching is done?

Thank you for your time and feedback on this.
Again, I'm no developer, so I know my code is probably not the best!

File: README.md
Comment (rnsc) [inline README.md:1]: I did a bit of linting on the README too, I hope that's ok

File: README.md
Comment (rnsc) [inline README.md:66]: I guess this example should be changed with the next version so that people are not confused?

File: README.md
Comment (rnsc) [inline README.md:27]: Same here, the verison of the action should be changed too right?

File: package.json
Comment (rnsc) [inline package.json:42]: I bumped it to a more recent version to have `getMultilineInput` support.

File: README.md
Comment (micalevisk) [reply README.md:27]: it should. `yarn version --no-git-tag-version --major`

File: README.md
Comment (micalevisk) [reply README.md:66]: right!

File: src/main.ts
Comment (micalevisk) [inline src/main.ts:104]: you might make this message more descriptive, or just drop this line idk

Comment (rnsc) [reply src/main.ts:104]: Indeed, I'll drop it, it's not needed, I just used it to debug the code logic

File: README.md
Comment (rnsc) [reply README.md:27]: right, @micalevisk, @Ana06, should we also change the repository URLs in the package.json? Or are we planning on trying to merge in the original repo in the future?

Comment (rnsc) [reply README.md:27]: I did the yarn command thingie, I'll let you all check :)

File: package.json
Comment (micalevisk) [inline package.json:59]: You could change this range to `^3.0.4`

**Review by micalevisk — CHANGES_REQUESTED** [review]

**Review by micalevisk — APPROVED** [review]

File: README.md
Comment (rnsc) [inline README.md:74]: @agraul not sure if you've seen this, but as GitHub actions core doesn't support a standard YAML as input, we have to declare the list a multi line string.

**Comment by rnsc** [comment]
Hello @Ana06
Do you think you'll have time this week ro review this PR again?

Thanks!

File: README.md
Comment (Ana06) [reply README.md:1]: That's awesome, thanks!

File: README.md
Comment (Ana06) [reply README.md:27]: should we call it `grob-filter` or just filter? I like `filter`, as we don't have any other filer I think the name doesn't need to be more specific as long as we document it properly. But I don't have a strong opinion... 🤔 :thinking: Any other opinion?

File: README.md
Comment (Ana06) [inline README.md:59]: Great section, thanks!

File: src/main.ts
Comment (Ana06) [inline src/main.ts:86]: should we be more explicit here and check that it starts by`!` instead of including it?

File: README.md
Comment (Ana06) [inline README.md:43]: I know this was already like this before, but I think it would be clearer to have this `Get all changed files as space-delimited` without the filter option and make another section to only get `*.php` and give there some more details about how the filter option works (that a list of regular expression is expected, GitHub like syntax is used, etc.). Ideally users should be able to understand how it works only with the README without checking the code. I am not completely sure this is the case now.

File: README.md
Comment (Ana06) [reply README.md:27]: @rnsc 
>should we also change the repository URLs in the package.json? Or are we planning on trying to merge in the original repo in the future? 

I think we can change the urls. I was hopping upstream integrates these changes when I originally forked it, but it doesn't look like it will happen.

**Review by Ana06 — COMMENTED** [review]
Awesome work! Just some minor comments. Once those are addressed/discussed, let's merge this and I can release version 2.0 🎉

File: README.md
Comment (rnsc) [reply README.md:44]: Added one example

File: src/main.ts
Comment (micalevisk) [inline src/main.ts:13]: minor typo here `gfilter`

**Review by micalevisk — CHANGES_REQUESTED** [review]

File: src/main.ts
Comment (rnsc) [reply src/main.ts]: yep, fixed

File: README.md
Comment (agraul) [reply README.md:84]: Thanks for the ping, I missed that. It's a bit unfortunate, I didn't know that lists can't be passed, but as long as the structure is kept I think this is still nice!

**Comment by rnsc** [comment]
Hello @Ana06 , I think I have made all the changes required, what would be the next steps? Not sure if can help still!

**Review by micalevisk — APPROVED** [review]

**Review by Ana06 — APPROVED** [review]
Great job @rnsc and sorry for the late response. I think it is ready to merge 🎉

File: .github/workflows/test.yml
Comment (Ana06) [inline .github/workflows/test.yml:89]: I had skipped that one 🙈 I have fix it on main 😉

Comment (rnsc) [reply .github/workflows/test.yml:89]: Oh no! Sorry!

**Pull request merged**
